build-tools/packages/build-cli/src/commands/generate/typetests.ts

```diff
@@ -0,0 +1,377 @@
+			default: false,
+		}),
+		...PackageCommand.flags,
+	} as const;
+
+	protected async processPackage(pkg: Package): Promise<void> {
+		const { outDir, outFile } = this.flags;
+
+		// This cast is safe because oclif has already ensured only known ApiLevel values get to this point.
+		const level = this.flags.level as ApiLevel;
+		const fallbackLevel = this.flags.publicFallback ? ApiLevel.public : undefined;
+
+		// Do not check that file exists before opening:
+		// Doing so is a time of use vs time of check issue so opening the file could fail anyway.
+		// Do not catch error from opening file since the default behavior is fine (exits process with error showing useful message)
+		const currentPackageJson = pkg.packageJson;
+		const previousPackageName = `${currentPackageJson.name}-previous`;
+		const previousBasePath = path.join(pkg.directory, "node_modules", previousPackageName);
+		const previousPackageJsonPath = path.join(previousBasePath, "package.json");
+
+		ensureDevDependencyExists(currentPackageJson, previousPackageName);
+		this.verbose(`Reading package.json at ${previousPackageJsonPath}`);
+		const previousPackageJson = (await readJson(previousPackageJsonPath)) as PackageJson;
+		// Set the name in the JSON to the calculated previous package name, since the name in the previous package.json is
+		// the same as current. This enables us to pass the package.json object to more general functions but ensure those
+		// functions use the correct name. For example, when we write the `import { foo } from <PACKAGE>/internal`
+		// statements into the type test file, we need to use the previous version name.
+		previousPackageJson.name = previousPackageName;
+
+		const typeTestOutputFile = getTypeTestFilePath(pkg, outDir, outFile);
+		if (currentPackageJson.typeValidation?.disabled === true) {
+			this.info("skipping type test generation because they are disabled in package.json");
+			rmSync(
+				typeTestOutputFile,
+				// force means to ignore the error if the file does not exist.
+				{ force: true },
+			);
+			this.verbose(`Deleted file: ${typeTestOutputFile}`);
+			this.exit(0);
+		}
+
+		const { typesPath: currentTypesPathRelative, levelUsed: currentPackageLevel } =
+			getTypesPathWithFallback(currentPackageJson, level, fallbackLevel);
+		const currentTypesPath = path.resolve(path.join(pkg.directory, currentTypesPathRelative));
+		this.verbose(
+			`Found ${currentPackageLevel} type definitions for ${currentPackageJson.name}: ${currentTypesPath}`,
+		);
+
+		const { typesPath: previousTypesPathRelative, levelUsed: previousPackageLevel } =
+			getTypesPathWithFallback(previousPackageJson, level, fallbackLevel);
+		const previousTypesPath = path.resolve(
+			path.join(previousBasePath, previousTypesPathRelative),
+		);
+		this.verbose(
+			`Found ${previousPackageLevel} type definitions for ${previousPackageJson.name}: ${previousTypesPath}`,
+		);
+
+		const currentFile = loadTypesSourceFile(currentTypesPath);
```

Comment:
I see that there was a subsequent change to use index.ts and assumes tsconfig.json. I think using the actual API is better and that we should build first. This also means that all of the packages that currently don't build tests separately would need updated to split, which is probably good anyway.
currentTypesPath is unused in main except for a verbose log call.